[DOC] Edits to Managing policies and permissions in Keycloak Authorization Services (#5298)
strimzi/strimzi-kafka-operator: documentation/modules/oauth/con-mapping-keycloak-authz-services-to-kafka-model.adoc

====

The `kafka` client definition must have the *Authorization Enabled* option enabled in the Keycloak Admin Console.

All permissions exist within the scope of the `kafka` client. If you have different Kafka clusters configured with different OAuth client IDs, they each need a separate set of permissions even though they're part of the same Keycloak realm.

When the Kafka client uses OAUTHBEARER authentication, the Keycloak authorizer (`KeycloakRBACAuthorizer`) uses the access token of the current session to retrieve a list of grants from the Keycloak server.
To retrieve the grants, the authorizer evaluates the Keycloak Authorization Services policies and permissions.

.Authorization scopes for Kafka permissions

An initial Keycloak configuration usually involves uploading authorization scopes to create a list of all possible actions that can be performed on each Kafka resource type.
This step is performed once only, before defining any permissions.
You can add authorization scopes manually instead of uploading them.

> Authorization scopes must contain all the possible Kafka permissions regardless of the resource type:

Reviewer: Just to clarify, using 'should' instead of 'must' was also good. Scopes that are not present can't be targeted, that's all. For example if 'IdempotentWrite' is not present, you can't create a rule matching IdempotentWrite action on a resource (e.g. some topic). By saying 'must' we just maybe reduce potential issues later when uses can't seem to target some permission because it's not listed under authorization scope. Thus, not listing all scopes will not result in any immediate errors. If you know what you're doing, for example that you'll never need IdempotentWrite, then you can as well omit it. But it's not bad to say 'must' either as I said - to minimise the need to fix authorization scopes list later on.

Author: Very helpful, thanks. I added a brief note:

> If you're certain you won't need a permission (for example, `IdempotentWrite`), you can omit it from the list of authorization scopes. However, that permission won't be available to target on Kafka resources.

[id="con-mapping-keycloak-authz-services-to-kafka-model_{context}"]

> = Mapping Keycloak Authorization Services to the Kafka authorization model

Reviewer 2: Are we using gerund on concepts?

[role="_abstract"]

> To grant Kafka permissions to user accounts or service accounts, use Keycloak Authorization Services rules on the OAuth client that represents the Kafka Broker.

Reviewer 2: I think we should introduce this section by saying how we do the _mapping_. That is, _we use the Kafka authorization model as a basis for defining the Keycloak roles and resources that will control access to Kafka..._

Author: That's great, thank you.

"The Kafka authorization model is used as a basis for defining the Keycloak roles and resources that will control access to Kafka."

> To grant Kafka permissions to user accounts or service accounts, use Keycloak Authorization Services rules on the OAuth client that represents the Kafka Broker.

Reviewer 2: ```suggestion
To grant Kafka permissions to user accounts or service accounts, you first create a _client_ specification for the Kafka broker in Keycloak. You then specify Keycloak Authorization Services rules on the client.
```

> Typically, the client id of the broker OAuth client is `kafka`.

Reviewer 2: In the example provided with Strimzi it is. Perhaps we should say that and that you should make sure the Kafka client is easy it identify.

Reviewer: You can use the same OAuth client ('kafka') with all your Kafka clusters. This gives you a single unified space of authorization rules. But you can also use different OAuth client ('my-cluster-kafka', 'cluster-dev-kafka' or something such) and define authorization rules specific to each cluster within each client configuration).

Author: Thanks. I summarised this in a note.

> Your OAuth 2.0 client definition for the Kafka Broker must have the *Authorization Enabled* option enabled.

Reviewer 2: ```suggestion
The `kafka` client definition must have the *Authorization Enabled* option enabled in the Keycloak Admin Console.
```

> All permissions exist within the scope of this OAuth 2.0 client. If you have different Kafka clusters configured with different OAuth 2.0 client IDs, they each need a separate set of permissions even though they're part of the same Keycloak realm.

Reviewer 2: ```suggestion
All permissions exist within the scope of the `kafka` client. If you have different Kafka clusters configured with different OAuth client IDs, they each need a separate set of permissions even though they're part of the same Keycloak realm.
```

Reviewer 2: I'm tending to think we don't use Oauth 2.0 everywhere. And just use OAuth. As long as we mention upfront in the OAuth section that it's OAuth 2.0?

Reviewer: Yeah, it's actually easier to read that way.

When the Kafka client uses OAUTHBEARER authentication, the Keycloak authorizer (`KeycloakRBACAuthorizer`) uses the access token of the current session to retrieve a list of grants from the Keycloak server.
To retrieve the grants, the authorizer evaluates the Keycloak Authorization Services policies and permissions.

> .About authorization scopes

Reviewer 2: ```suggestion
.Authorization scopes for Kafka permissions
```

An initial Keycloak configuration usually involves uploading authorization scopes to create a list of all possible actions that can be performed on each Kafka resource type.
This step is performed once only, before defining any permissions.
You can add authorization scopes manually instead of uploading them.

> Authorization scopes should contain all the possible Kafka permissions regardless of the resource type:

Reviewer 2: ```suggestion
Authorization scopes must contain all the possible Kafka permissions regardless of the resource type:
```

* `Create`
* `Write`
* `Read`
* `Delete`
* `Describe`
* `Alter`
* `DescribeConfig`
* `AlterConfig`
* `ClusterAction`
* `IdempotentWrite`

> .Defining resource patterns for permissions checks

Reviewer 2: ```suggestion
.Resource patterns for permissions checks
```

> The resources use pattern names for pattern matching against the targeted resources when performing permission checks.

Reviewer 2: ```suggestion
Resource patterns are used for pattern matching against the targeted resources when performing permission checks.
```

> The general pattern is `__RESOURCE-TYPE:PATTERN-NAME__`.

Reviewer 2: ```suggestion
The general pattern format is `__RESOURCE-TYPE:PATTERN-NAME__`.
```

> The resource types mirror the Kafka authorization model.

Reviewer 2: An example might be useful here.

Author: I'm going to resolve this one.

The pattern allows for two matching options:

> * Exact matching (when the pattern does not end with `\*`)

Author: @mstruk - Is `\*` a typo? Should it say does not end with * ?

Reviewer: Looks to me like an escape char for Markdown. The text should display only asterisk.

* Prefix matching (when the pattern ends with `*`)

.Example patterns for resources
[source]
----
Topic:my-topic
Topic:orders-*
Group:orders-*
Cluster:*
----

> Additionally, the general pattern can be prefixed by `kafka-cluster:__CLUSTER-NAME__` followed by a comma, where _CLUSTER-NAME_ refers to the `metadata.name` in the Kafka custom resource.

Reviewer 2: ```suggestion
Additionally, the general pattern format can be prefixed by `kafka-cluster:__CLUSTER-NAME__` followed by a comma, where _CLUSTER-NAME_ refers to the `metadata.name` in the Kafka custom resource.
```

Reviewer 2: This looks like a different pattern entirely, unless I'm mistaken?

.Example patterns for resources with cluster prefix
[source]
----
kafka-cluster:my-cluster,Topic:*
kafka-cluster:*,Group:b_*
----

When the `kafka-cluster` prefix is missing, it is assumed to be `kafka-cluster:*`.

When defining a resource, you can associate it with a list of possible authorization scopes which are relevant to the resource.
Set whatever actions make sense for the targeted resource type.

Though you may add any authorization scope to any resource, only the scopes supported by the resource type are considered for access control.

> .About policies

Reviewer 2: ```suggestion
.Policies for user and service access
```

Policies are used to target permissions to one or more user accounts or service accounts.
Targeting can refer to:

* Specific user or service accounts
* Realm roles or client roles
* User groups
* JavaScript rules to match a client IP address

A policy is given a unique name and can be reused to target multiple permissions to multiple resources.

> .Defining permissions based on scopes, resources, and policies

Reviewer 2: ```suggestion
.Permissions to grant access
```

Use fine-grained permissions to pull together the policies, resources, and authorization scopes that grant access to users.

> The name of each permission should clearly define which permissions it grants to which users.

Reviewer: I would add an example of permission name here.

Author: Something like this? `Dev Team B read from topics starting with x`.

Reviewer: Yes, that looks good.
